active_storage/server/app.py

```diff
@@ -1,162 +1,215 @@
-from fastapi import FastAPI, Depends, Path, Request
-from fastapi.exceptions import RequestValidationError
-from fastapi.responses import Response, StreamingResponse
-
-import httpx
+from fastapi import FastAPI, HTTPException, Depends
+from fastapi.responses import Response
+from fastapi.security import HTTPBasic, HTTPBasicCredentials
 
+import itertools
+import botocore
+import aioboto3
 import numpy as np
 
-from .config import settings
-from .models import NumericDataType
+from .models import RequestData, Reducer, AllowedDatatypes, ALLOWED_OPERATIONS
 
 
 app = FastAPI()
+security = HTTPBasic()
 
 
-@app.on_event("startup")
-def on_startup():
-    """
-    Initialises the HTTPX client on startup.
-    """
-    app.state.client = httpx.AsyncClient()
-
-
-@app.on_event("shutdown")
-async def on_shutdown():
-    """
-    Closes the HTTPX client on shutdown.
-    """
-    await app.state.client.aclose()
+class S3Exception(Exception):
+    """ Custom exception class for catching upstream S3 exception """
+    def __init__(self, upstream_response):
+        super().__init__(self)
+        self.upstream_response = upstream_response
 
 
-@app.exception_handler(RequestValidationError)
-async def handle_validation_exception(request, exc):
+@app.exception_handler(S3Exception)
+async def handle_upstream_s3_exception(request, exc: S3Exception):
```

Comment:
We should make this a JSON response. It is only XML in order to comply with the S3 interface in the previous implementation.